src/ctap/apdu.rs

```diff
@@ -0,0 +1,303 @@
+use alloc::vec::Vec;
+use core::convert::TryFrom;
+
+type ByteArray = &'static [u8];
+
+#[cfg_attr(test, derive(Clone, Debug))]
+#[allow(non_camel_case_types)]
+#[derive(PartialEq)]
+pub enum ApduStatusCode {
+    SW_SUCCESS,
+    /// Command successfully executed; 'XX' bytes of data are
+    /// available and can be requested using GET RESPONSE.
+    SW_GET_RESPONSE,
+    SW_WRONG_DATA,
+    SW_WRONG_LENGTH,
+    SW_COND_USE_NOT_SATISFIED,
+    SW_FILE_NOT_FOUND,
+    SW_INCORRECT_P1P2,
+    /// Instruction code not supported or invalid
+    SW_INS_INVALID,
+    SW_CLA_INVALID,
+    SW_INTERNAL_EXCEPTION,
+}
+
+impl From<ApduStatusCode> for ByteArray {
+    fn from(status_code: ApduStatusCode) -> ByteArray {
+        match status_code {
+            ApduStatusCode::SW_SUCCESS => b"\x90\x00",
+            ApduStatusCode::SW_GET_RESPONSE => b"\x61\x00",
+            ApduStatusCode::SW_WRONG_DATA => b"\x6A\x80",
+            ApduStatusCode::SW_WRONG_LENGTH => b"\x67\x00",
+            ApduStatusCode::SW_COND_USE_NOT_SATISFIED => b"\x69\x85",
+            ApduStatusCode::SW_FILE_NOT_FOUND => b"\x6a\x82",
+            ApduStatusCode::SW_INCORRECT_P1P2 => b"\x6a\x86",
+            ApduStatusCode::SW_INS_INVALID => b"\x6d\x00",
+            ApduStatusCode::SW_CLA_INVALID => b"\x6e\x00",
+            ApduStatusCode::SW_INTERNAL_EXCEPTION => b"\x6f\x00",
+        }
+    }
+}
+
+#[allow(non_camel_case_types, dead_code)]
+pub enum ApduIns {
+    SELECT = 0xA4,
+    READ_BINARY = 0xB0,
+    GET_RESPONSE = 0xC0,
+}
+
+#[cfg_attr(test, derive(Clone, Debug))]
+#[allow(dead_code)]
+#[derive(Default, PartialEq)]
+pub struct ApduHeader {
+    cla: u8,
+    ins: u8,
+    p1: u8,
+    p2: u8,
+}
+
+impl From<&[u8]> for ApduHeader {
+    fn from(header: &[u8]) -> Self {
+        ApduHeader {
+            cla: header[0],
+            ins: header[1],
+            p1: header[2],
+            p2: header[3],
+        }
+    }
+}
+
+#[cfg_attr(test, derive(Clone, Debug))]
+#[allow(dead_code)]
+#[derive(Default, PartialEq)]
+pub struct APDU {
+    header: ApduHeader,
+    lc: u16,
```

Comment:
Is `lc` actually needed once parsed? It's supposed to be equivalent to `data.len()` at the end otherwise the APDU is ill-formed, isn't it?

Reply:
So I shouldn't keep it even if it is to keep the structure of what the actual APDU looked like?

Reply:
Actually the code currently doesn't take `Lc` into account. So it's fine to keep it.
We'll add sanity checks later.